lib/std/time.zig

```diff
@@ -344,6 +344,221 @@ test Timer {
     try testing.expect(timer.read() < time_1);
 }
 
+pub fn Time(precision_: comptime_int, comptime has_offset: bool) type {
+    const multiplier: comptime_int = try std.math.powi(usize, 10, precision_);
+    return struct {
+        hour: Hour = 0,
+        minute: Minute = 0,
+        /// Allows leap seconds.
+        second: Second = 0,
+        /// Milliseconds, microseconds, or nanoseconds.
+        subsecond: Subsecond = 0,
+        /// Offset of `hour` and `minute` from in minutes from UTC.
+        offset: IMinutes = 0,
+
+        pub const Hour = IntFittingRange(0, 23);
```

Comment:
might be a bit wild but what about making the range check for the date(time) fields a debug feature, which you could disable in a "fast" build?

Reply:
I already wrote range checks in `toRfc3339` functions. I can refactor that out into `fn validate(self) !void` functions.

Where would you like these range checks?

Reply:
That's a partly dependent on the rest of the implementation, but I think the only place you need the check is when you create the Date(Time) instance from the fields. In zdt, I have a [fromFields](https://codeberg.org/FObersteiner/zdt/src/branch/main/lib/Datetime.zig#L142) method, which is the place where I naturally validate the fields. Anyways, this was just an idea. I'm not sure field validation will turn into a bottle-neck, and I don't want this to be a premature optimization thing ;-)

Reply:
Integer field construction makes sense if there are good default `Date`, `Time`, and `DateTime` validators for each calendar system. However, I don't think any of these exist.

Real life `Date`s are validated based on calendar system and political government. While validating Gregorian date month+days given the year sounds good (and I have implemented it for RFC 3339), there's still the political government problem of when it was adopted. Unix `cal` claims 1752:
```sh
$ cal Sept 1752
   September 1752
Su Mo Tu We Th Fr Sa
       1  2 14 15 16
17 18 19 20 21 22 23
24 25 26 27 28 29 30
```

Is 1752-11-03 a valid Gregorian calendar date? I don't have the authority to give that answer, I'd rather let the user decide.
 
`Time` and `DateTime` validation is also difficult. Some time systems consider `24:00:00` a valid time that represents the instant at the end of the day, others don't. Handling leap seconds for times like `1990-12-31T15:59:60-08:00` requires:
1. The time know about the date, and
2. A leap second table, which incurs a performance penalty.

For this reason I don't think integer field construction nor validation outside of RFCs/ISOs belong in the stdlib.

Reply:
seems I didn't get your intention when using IntFittingRange. To me, the feels like a question of how opinionated the implementation should get. But I don't think this comment section is the right place to have this discussion - let's call this resolved here.

Reply:
`IntFittingRange` doesn't and can't do range checking. It's just a placeholder for the smallest integer type that will fit. Though, it is true that using non-native sizes incurs a performance penalty right now, even when it shouldn't. A function like this could work:

```zig
fn NativeInt(comptime Int: type) type {
    const ti = @typeInfo(Int).Int;
    return @Type(.{ .Int = .{
        .bits = std.math.ceilPowerOfTwo(u65535, ti.bits) catch unreachable,
        .signedness = ti.signedness,
    } });
}```

Reply:
IntFittingRange can't and doesn't do any range checking. It's just a shortcut for the smallest int that will fit. However, using non-native integer sizes incurs a performance penalty right now, even when it shouldn't.

Reply:
@FObersteiner My intention with `IntFittingRange` is to pick the smallest required type while documenting the _expected_ but _not enforced_ range of values.

@notcancername I did not know non powers of two incurs a performance penalty. Can you link the relevant issue(s)?

Reply:
I don't know whether there's an issue for it, but this code demonstrates the problem quite nicely.

https://godbolt.org/z/MTrvzz4nd

Reply:
I opened #19616 .